src/third_party/libkmip-0ecda33/kmippp/kmippp.cpp

```diff
@@ -153,8 +154,132 @@ cert_chain read_certificate_chain(const std::string& client_cert_fn,
+        case KMIP_REASON_INVALID_ATTRIBUTE:                      return "Invalid Attribute";
+        case KMIP_REASON_INVALID_ATTRIBUTE_VALUE:                return "Invalid Attribute Value";
+        case KMIP_REASON_INVALID_CORRELATION_VALUE:              return "Invalid Correlation Value";
+        case KMIP_REASON_INVALID_CSR:                            return "Invalid CSR";
+        case KMIP_REASON_INVALID_OBJECT_TYPE:                    return "Invalid Object Type";
+        case KMIP_REASON_KEY_WRAP_TYPE_NOT_SUPPORTED:            return "Key Wrap Type Not Supported";
+        case KMIP_REASON_MISSING_INITIALIZATION_VECTOR:          return "Missing Initialization Vector";
+        case KMIP_REASON_NON_UNIQUE_NAME_ATTRIBUTE:              return "Non Unique Name Attribute";
+        case KMIP_REASON_OBJECT_DESTROYED:                       return "Object Destroyed";
+        case KMIP_REASON_OBJECT_NOT_FOUND:                       return "Object Not Found";
+        case KMIP_REASON_NOT_AUTHORISED:                         return "Not Authorised";
+        case KMIP_REASON_SERVER_LIMIT_EXCEEDED:                  return "Server Limit Exceeded";
+        case KMIP_REASON_UNKNOWN_ENUMERATION:                    return "Unknown Enumeration";
+        case KMIP_REASON_UNKNOWN_MESSAGE_EXTENSION:              return "Unknown Message Extension";
+        case KMIP_REASON_UNKNOWN_TAG:                            return "Unknown Tag";
+        case KMIP_REASON_UNSUPPORTED_CRYPTOGRAPHIC_PARAMETERS:   return "Unsupported Cryptographic Parameters";
+        case KMIP_REASON_UNSUPPORTED_PROTOCOL_VERSION:           return "Unsupported Protocol Version";
+        case KMIP_REASON_WRAPPING_OBJECT_ARCHIVED:               return "Wrapping Object Archived";
+        case KMIP_REASON_WRAPPING_OBJECT_DESTROYED:              return "Wrapping Object Destroyed";
+        case KMIP_REASON_WRAPPING_OBJECT_NOT_FOUND:              return "Wrapping Object Not Found";
+        case KMIP_REASON_WRONG_KEY_LIFECYCLE_STATE:              return "Wrong Key Lifecycle State";
+        case KMIP_REASON_PROTECTION_STORAGE_UNAVAILABLE:         return "Protection Storage Unavailable";
+        case KMIP_REASON_PKCS11_CODEC_ERROR:                     return "PKCS#11 Codec Error";
+        case KMIP_REASON_PKCS11_INVALID_FUNCTION:                return "PKCS#11 Invalid Function";
+        case KMIP_REASON_PKCS11_INVALID_INTERFACE:               return "PKCS#11 Invalid Interface";
+        case KMIP_REASON_PRIVATE_PROTECTION_STORAGE_UNAVAILABLE: return "Private Protection Storage Unavailable";
+        case KMIP_REASON_PUBLIC_PROTECTION_STORAGE_UNAVAILABLE:  return "Public Protection Storage Unavailable";
+    }
+    return "Unknown";
+}
+
+std::string generate_error_message(int status, const LastResult* last_result) {
+    if (status < 0) {
+        // @see the `KMIP_<smth>` constants in the beginning
+        // of the `kmip.h` file
```

Comment:
The `result_reason` enum may change in one of the three ways:
1. removing some enumerator (e.g. `KMIP_REASON_PKCS11_CODEC_ERROR`);
2. adding another enumerator
3. changing the value of existing enumerators

Case 3 is the simplest to analyze. Since the `generate_error_message` function checks all enumerators in the `switch` statement, its result is not affected by any number of enumerators changing their values.

In case 1, the removed enumerator becomes undefined and the compilation will fail.

Case 2 is the most interesting. I tried to remove the `KMIP_REASON_PKCS11_CODEC_ERROR` from the switch statement pretending the upstream `libkmip` added an enumerator and `generate_error_message` failed to reflect that. As a result, I got the following compilation error:
```
src/third_party/libkmip-0ecda33/kmippp/kmippp.cpp: In function 'const char* kmippp::{anonymous}::kmip_reason_to_str(result_reason)':
src/third_party/libkmip-0ecda33/kmippp/kmippp.cpp:159:12: warning: enumeration value 'KMIP_REASON_PKCS11_CODEC_ERROR' not handled in switch [-Wswitch]
  159 |     switch (reason) {
      |            ^
```
It makes me think that the compiler treats the absence of the `default` label in the `switch` statement as an intention to handle all the enumerators and, given the required compilation options are in place (`-Wswitch`), notifies if that hasn't been done.

As a result of the analysis above, we may conclude that the `generate_error_message` is safe against any combination of changes to the upstream `result_reason` enumeration.

Reply:
Well, now I see we have two things here:
- switch statement in the `kmip_reason_to_str` function (~50 cases)
- `errors` array in `generate_error_message` function (21 elements)
The switch statement looks safe.
Array is indexed by status value. Status values are just #defines from KMIP_NOT_IMPLEMENTED=-1 to KMIP_INVALID_LENGTH= -21. If any of those defines is changes then `errors` array is out of sync.
I understand such changes are unlikely but if we want more robust code we need to replace `errors` array with another switch statement.

Reply:
Since we do not plan to change any KMIP_xxxx defines I approve this PR.